models/utils.py

```diff
@@ -1,5 +1,93 @@
 import tensorflow as tf
 import mesh_tensorflow as mtf
+from functools import partial
+
+def entmax_backward(explicit_inputs, all_inputs, forward_operations, outputs, output_grads, alpha = 1.3, dim = None, n_iter = 50):
+    x, = explicit_inputs
+    y, = outputs
+    dY, = output_grads
+
+    gppr = mtf.where(mtf.greater(y, 0), mtf.pow(y, (2 - alpha)), mtf.zeros_like(y))
+    dX = dY * gppr
+
+    q = mtf.reduce_sum(dX, reduced_dim = dim) / mtf.reduce_sum(gppr, reduced_dim = dim)
+    dX = dX - q * gppr
+
+    return dX,
+
+def entmax_forward(x, alpha = 1.3, dim = None, n_iter = 50):
+    assert alpha > 1 and alpha < 2, 'alpha must be between 1 and 2'
+
+    _gp = lambda x, alpha: x ** (alpha - 1)
+    _gp_inv = lambda x, alpha: mtf.pow(x, (1 / (alpha - 1)))
+    _p = lambda x, alpha: _gp_inv(mtf.relu(x), alpha)
+
+    dim = x.shape[-1] if dim is None else dim
+    d = dim.size
+
+    x = x * (alpha - 1)
+
+    max_val = mtf.reduce_max(x, reduced_dim = dim)
+
+    tau_lo = max_val - _gp(1, alpha)
+    tau_hi = max_val - _gp(1 / d, alpha)
+
+    f_lo = mtf.reduce_sum(_p(x - tau_lo, alpha), reduced_dim = dim) - 1
+
+    dm = tau_hi - tau_lo
+
+    for _ in range(n_iter):
```

Comment:
am i supposed to be using `mtf.while_loop` here?

Reply:
I think that a for loop is fine here? 
Using mtf.assign() might be more efficient than dm = dm/2, since a variable will get added to the graph for each iteration in the for loop here. Probably not a problem if n_iters stays around 50 though.